Source/WebCore/loader/MixedContentChecker.cpp

```diff
@@ -125,6 +136,56 @@ bool MixedContentChecker::frameAndAncestorsCanRunInsecureContent(LocalFrame& fra
     return allowed;
 }
 
+bool MixedContentChecker::shouldUpgradeInsecureContent(LocalFrame& frame, IsUpgradable isUpgradable, const URL& url, FetchOptions::Mode mode, FetchOptions::Destination destination, Initiator initiator)
+{
+    RefPtr document = frame.document();
+    if (!document->settings().upgradeMixedContentEnabled() || isUpgradable != IsUpgradable::Yes)
+        return false;
+
+    // The request should not be upgraded if:
+    // 4.1.3 § 4.3 Does settings prohibit mixed security contexts? returns "Does Not Restrict Mixed Security Contents" when applied to request’s client.
+    if (!foundMixedContentInFrameTree(frame, url))
+        return false;
+
+    auto shouldUpgradeIPAddress = document->settings().iPAddressMixedContentUpgradeTestingEnabled();
+
+    // The request's URL is not upgraded in the following cases.
+    // 4.1.1 request’s URL is a potentially trustworthy URL.
+    if (url.protocolIs("https"_s)
+        // 4.1.2 request’s URL’s host is an IP address.
+        || (!shouldUpgradeIPAddress && URL::hostIsIPAddress(url.host()))
+        // 4.1.4 request’s destination is not "image", "audio", or "video".
+        || (destination != FetchOptions::Destination::Audio && destination != FetchOptions::Destination::Image && destination != FetchOptions::Destination::Video)
+        // 4.1.5 request’s destination is "image" and request’s initiator is "imageset".
+        || (destination == FetchOptions::Destination::Image && initiator == Initiator::Imageset)
+        // and CORS is excluded
+        || mode == FetchOptions::Mode::Cors)
+        return false;
+    return true;
+}
+
+bool MixedContentChecker::shouldBlockInsecureContent(LocalFrame& frame, IsUpgradable isUpgradable, const URL& url, FetchOptions::Mode mode, FetchOptions::Destination destination, Initiator initiator)
+{
+    RefPtr document = frame.document();
+    if (!document->settings().upgradeMixedContentEnabled())
+        return false;
+    if (!foundMixedContentInFrameTree(frame, url))
+        return false;
+
+    if (isUpgradable == IsUpgradable::Yes) {
+        if (frame.settings().allowDisplayOfInsecureContent())
```

Comment:
allowDisplayOfInsecureContent or allowRunningOfInsecureContent, or should it be a parameter?

Reply:
"runnable" content should not be upgradable, so I don't think we need to check that? This is similar to how frameAndAncestorsCanDisplayInsecureContent doesn't check allowRunningOfInsecureContent.